app/services/google/parser_service.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module Google
+  class ParserService
+    NON_ADS_RESULT_SELECTOR = 'a[data-ved]:not([role]):not([jsaction]):not(.adwords):not(.footer-links)'
+    AD_CONTAINER_ID = 'tads'
+    ADWORDS_CLASS = 'adwords'
+
+    def initialize(html_response:)
+      raise ArgumentError, 'response.body cannot be blank' if html_response.body.blank?
+
+      @html = html_response
+
+      @document = Nokogiri::HTML.parse(html_response)
+
+      # Add a class to all AdWords link for easier manipulation
+      @document.css('div[data-text-ad] a[data-ved]').add_class(ADWORDS_CLASS)
+
+      # Mark footer links to identify them
+      @document.css('#footcnt a').add_class('footer-links')
+    end
+
+    # Parse html data and return a hash with the results
+    def call
+      {
+        ads_top_count: ads_top_count,
+        ads_page_count: ads_page_count,
+        ads_top_url: ads_top_url,
+        ads_page_url: ads_page_url,
+        non_ads_result_count: non_ads_result_count,
+        non_ads_url: non_ads_url,
+        total_link_count: total_link_count,
+        html: @html
+      }
+    end
+
+    private
```

Comment:
Related to https://github.com/malparty/google-search-ruby/pull/46#discussion_r660288299, I think we can do here also, defining `attr_reader` and use accessor while retrieving value
```suggestion
    private
    
    attr_reader :html, document
```

Reply:
True, nice catch! 
Done ✅